docs/confluence/advanced/grounding.md

```diff
@@ -0,0 +1,154 @@
+# Grounding 
+
+## Flag
+
+Grounding is on by default. To turn it off, use `--settings -smt_groundQuantifiers=false`
+
+ 
+## Grounding General Idea
+
+Grounding should make using quantifiers much easier on the solvers. For example,
+
+```
+assume forall x. f(x) = 0
+```
+
+A very natural init state axiom. Instead of giving it as is to the SMT-solvers, we automatically collect all instances of `f` in the formula we want to verify, e.g., `f(2), f(9), f(y + 3), f(z)`, and replace the original assert with:
+
+```
+assume f(2) = 0
+assume f(9) = 0
+assume f(y + 3) = 0
+assume f(z) = 0
+```
+
+Simple enough! You can of course write more complex quantified statements, and they will be grounded in a similar fashion. However, this approach has its limitations. It will never say a rule is valid when it is not (i.e., grounding is sound), but it may generate false counter-examples. Read below on how to minimize the chance of this happening. 
+
+Note that this works not only for assumes, but also for asserts, axioms, invariants, and anywhere else quantifiers are used. You can also use `exists` quantifiers.
+
+I’ve seen examples where this changes a timeout to 3 seconds, cases where it didn’t really change the running time, and a rare case where it performed much worse than standard quantification. but in general, this should work much better in terms of running time. 
+
+ 
+## Limitations
+
+There are some cases we just can’t ground, some cases that are incomplete (i.e., give a wrong counter-example), but grounding is always sound (i.e., if we say a formula is valid then it surely is). Here are some guidelines and explanations.
+
+ 
+### Alternating Quantifiers
+
+```
+assume forall x, exists y. f(x) = g(y)
+```
+
+
+Doesn’t currently work. We can make it work, so if you need it, give us a shout and we’ll make an effort :slight_smile: 
+
+```
+assume exists x, forall y ...
+```
+
+Does work! and also:
+
+```
+assert forall x, exists y.
+```
+
+For the first case,  we currently crash the tool. In such a case, preferably rewrite your spec rerun, or you can turn grounding off and see if the solvers manage it.
+
+ 
+### Recursion
+
+```
+assume forall x. f(x) > f(x+1)
+assert f(8) > f(6)
+```
+
+Any sort of such recursion will very likely give a wrong counter-example. We do not auto-detect it, so the responsibility to avoid it and to realize the counter-example is wrong is on the user.
+
+ 
+### Variables That Are Not Arguments
+
+```
+forall x. x^2 != y
+```
+
+Can’t work, because for grounding we run over all instances of functions in the formula. Here there is no function and no instance. This won’t work either:
```

Comment:
This cannot work, as we run over all instances of functions in the formula during grounding. In this case,there is no function and instance and therefore it would not work either: